### ISSUE TITLE
Osu lazer runs in background on ubuntu 16.04

### ISSUE THREAD

**Issue by siddharthkhamithkar** [ppy/osu#1834]
After updating to the latest release, running mono ./osu\!.exe runs the game in the background. It doesn't launch.
![screenshot from 2018-01-02 20-10-39](https://user-images.githubusercontent.com/31732674/34487152-0ed48498-eff9-11e7-97c2-5fac5d083efc.png)
![screenshot from 2018-01-02 20-11-48](https://user-images.githubusercontent.com/31732674/34487179-2d96cf26-eff9-11e7-9cb8-600c5c53d264.png)

**Comment by TheElectronWill** [comment]
I've just got the exact same problem on Fedora 26!
Infos:
```
Mono JIT compiler version 5.4.1.6 (tarball Wed Nov  8 16:59:44 UTC 2017)
NuGet Version: 4.4.1.4656
Linux Kernel 4.14.8 x86_64
```

**Comment by peppy** [comment]
Logs? Tried to debug?

**Comment by TheElectronWill** [comment]
Here you go: [runtime.log](https://github.com/ppy/osu/files/1599411/runtime.log)
Running `mono --debug ./osu\!.exe` doesn't output anything more...
As I'm on a laptop with Intel+Nvidia, I tried to launch osu! with both the intel chipset and the nvidia gpu, but that didn't change anything.

**Comment by peppy** [comment]
I mean debugging using an IDE and stuff.

**Comment by TheElectronWill** [comment]
Just a few minutes to install a decent C# IDE and I'll do that.

**Comment by TheElectronWill** [comment]
Allright, so this is definitely a weird bug:
- Running from the Monodevelop IDE (with `Run` action) makes osu!lazer work perfectly
- Running from the Terminal with `mono <exe file>` doesn't spawn any window and takes a lot of CPU. 
- Debugging the same exe file from the IDE makes it work.

With `strace mono <exe file>`  I've got the following calls trace: [strace.log](https://github.com/ppy/osu/files/1599475/strace.log). The last line repeats forever while the program eats all the CPU.
 ```
mmap(0x40e45000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x40e45000
recvmsg(6, {msg_namelen=0}, 0)          = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(6, {msg_namelen=0}, 0)          = -1 EAGAIN (Resource temporarily unavailable)
...
```
Reproducible with Mono 5.4.1 Stable and Mono 5.8.0 Beta

**Comment by TheElectronWill** [comment]
After more investigation I've discovered something interesting.

The program hangs in the method DesktopGameWindow.windowMode_ValueChanged [on line 134](https://github.com/ppy/osu-framework/blob/6134dafccb3368dac96d837537325c04b89fb8ee/osu.Framework/Platform/DesktopGameWindow.cs#L134).
I decomposed `Position.set` to isolate the problem:
```cs
int displayWidth = DisplayDevice.Default.Width;
int displayHeight = DisplayDevice.Default.Height;
// A log message put here always appears so it's fine
int drawingWidth = Size.Width;
int drawingHeight = Size.Height;
// Fine too
Point p = new Point(
    (int)Math.Round((displayWidth - drawingWidth) * value.X),
    (int)Math.Round((displayHeight - drawingHeight) * value.Y));
// Still fine
Location = p; // The problem is here! mono hangs indefinitely in this call.
// Code here is never executed
```
Therefore the guilty code lies inside OpenTK's [Location.set](https://github.com/ppy/opentk/blob/905c1506391af568874f85a22151634ed03ccfee/src/OpenTK/NativeWindow.cs#L322), in the call to `EnsureUndisposed`.

Running `mono --trace=all <osu exe>` shows that the process is stuck on some X11-related resource. Here are the last 250 lines - didn't put more because everything repeats endlessly.
[with_mono_cmd_trace=all_last250.log](https://github.com/ppy/osu/files/1602266/with_mono_cmd_trace.all_last250.log)

**So it could be a bug in OpenTK's Linux implementation** (your fork or the official upstream?).

Regarding the fact that it works on the IDE, my (totally unsure and unproven) guess is that the bug might be caused by the lack of a parent window or process, which might be provided by the IDE but not by calling mono from the terminal.

**Comment by peppy** [comment]
Out of curiosity, did you have lazer running previously? We recently updated opentk, so I'm wondering if a previous build was working or not (this may have been a regression)

**Comment by TheElectronWill** [comment]
I haven't tried any previous build but I could do so ^^

**Comment by peppy** [comment]
One thing to note is that we were using a nightly opentk package which is no longer available at the nuget source (myget in this case) so it may be a bit annoying to test.

**Comment by VPeruS** [comment]
@TheElectronWill program stuck in implementation.Location setter, in this call https://github.com/opentk/opentk/blob/254f7e9bfb2627e0cdedc61f4cb0e1f824f11f56/src/OpenTK/Platform/X11/X11GLNative.cs#L1258. If you look at it you will spot https://github.com/opentk/opentk/blob/254f7e9bfb2627e0cdedc61f4cb0e1f824f11f56/src/OpenTK/Platform/X11/X11GLNative.cs#L834 this is what cause error, "if" check specific event, if not -> continue wait. But it never check for ConfigureNotify(but function was called exclusively for this event)

**Comment by Frassle** [comment]
That event is covered by XCheckWindowEvent because it's a window event.

I'd guess what's happening is that your calling Location set but your not actually changing the size or position. So OpenTK never calls any of the XMove/Resize calls but still calls ProcessEvents and waits for the event. Should be a simple fix of just adding an "else { return }" to the if-chain in Bounds.set which checks if size or position are different,

**Comment by IsoaSFlus** [comment]
I got the same problem on ubuntu 17.10.
I compiled osu by [AUR](https://aur.archlinux.org/packages/osu-lazer-git/) on archlinux. It works well on Archlinux. 
But when I move the binary to my ubuntu 17.10, I got the same problem.

**Comment by ShadowPower** [comment]
The same problem on linux mint 18.3 KDE.
I compile osu!lazer on Windows, because I got some problems while compiling in mint.
Works fine in Windows but Linux.

**Comment by peppy** [comment]
Please stop posting these "happens to me" posts. It's already been established that an issue exists, and a solution has graciously been offered by @Frassle. If you have the time and expertise, please try this solution (and report back or PR if it works), else wait for someone to do this.

**Comment by TemTemmie** [comment]
try reinstalling mono or building from source.
if not working , try compiling osu using [wine](https://winehq.org)

**Comment by gtensha** [comment]
I can confirm that the solution offered by @Frassle works on my machine. Using Debian testing, kernel 4.14.7 and mono 5.8.0.108. Compiling OpenTK (from ppy/opentk) via the build.sh script did fail the window resize test, however: 

`OpenTK.Tests.Integration.GameWindow+Borders.Can resize hidden borders [FAIL]`

No issues in game as far as I can tell, and resizing the window works as normal. I should add that I do not have sufficient experience with C# or OpenTK for that matter, and I don't know what should be considered a valid solution. Should I make a pull request to ppy/opentk for this? It seems like a hacky solution, considering the failed test and all. However, that test fails even without the change on my machine so it should be an improvement regardless (assuming the issue doesn't exist solely on my machine.)

**Comment by VPeruS** [comment]
failed test isn't important(it always fails on linux), make pr to opentk/opentk

**Comment by gtensha** [comment]
I'm sorry, it seems I got a little ahead of myself. The issue isn't there at all anymore, I just tested with a fresh clone of master, downloading all the nuget packages over again. I ran nuget restore in the first place, but it seems like it kept using the old OpenTK version (the package was still there.) As such, the solution the first time was simply downloading and building the latest osu-compatible OpenTK from ppy/opentk, not my new changes. I can't test with the newest version from the official repository, because Lazer does not build using it. I'm sorry for my incompetence, but if someone else can test it themselves and also confirm it as working then I guess the issue no longer remains.
